File: cli/commands/commit.py
# CLI command to generate and make a commit using the configured provider
from core.config import settings
from providers.factory import get_provider
from utils import get_branch, get_diff, stage_all, commit, push, clean_commit_message, Colors

def get_ticket_prefix(branch):
	import re
	m = re.match(r'^([A-Za-z]+-\d+)', branch)
	return m.group(1) if m else ''

def main():
	import argparse
	parser = argparse.ArgumentParser(description="Generate and commit using AI")
	parser.add_argument('--format', choices=['detailed', 'one-line'], help='Commit message format')
	parser.add_argument('--provider', help='Provider to use (overrides config)')
	parser.add_argument('--model', help='Model to use (overrides config)')
	parser.add_argument('--push', action='store_true', help='Push after commit')
	args = parser.parse_args()

	if args.format:
		settings.set('COMMIT_FORMAT', args.format)
	provider_name = args.provider or settings.get('PROVIDER', 'openai')
	provider_kwargs = {}
	if args.model:
		provider_kwargs['model'] = args.model
	provider = get_provider(provider_name, **provider_kwargs)

	stage_all()
	diff = get_diff(staged=True)
	if not diff.strip():
		print(Colors.info("ℹ No staged changes to commit."))
		return
	branch = get_branch()
	prefix = get_ticket_prefix(branch)
	short = settings.get('COMMIT_FORMAT', 'detailed') == 'one-line'
	sys_msg = (
		"You are an expert Git commit assistant. When responding, return only the commit message text itself—no extra explanation, quotes, or formatting. "
		"First, inspect the current Git branch name. If it begins with a ticket code matching the pattern LETTERS-DIGITS (for example ABC-123 or EL-2024), capture that exact code and place it at the very start of your message, followed by a colon and a space. If no ticket code is present, do not include any prefix. "
		"Next, identify the primary change or task implied by the branch name and present it as the first action in your commit message. Then, describe any secondary updates, fixes, or refactoring included in this commit. "
		"Use a natural, professional tone that reads like a teammate clearly explaining the work you’ve done. Use bullet points to separate multiple actions if they exist."
	)
	user_msg = f"Branch: {branch}\nWrite a {'one-line' if short else 'detailed, human-friendly'} commit message for these changes:\n\n{diff}"
	
	print(Colors.header("🤖 Generating commit message with AI..."))
	commit_msg = provider.generate(
		prompt=user_msg,
		messages=[
			{"role": "system", "content": sys_msg},
			{"role": "user", "content": user_msg}
		]
	)
	
	# Clean up the AI-generated commit message
	commit_msg = clean_commit_message(commit_msg)
	
	if prefix and not commit_msg.startswith(prefix):
		commit_msg = f"{prefix}: {commit_msg}"
	
	print(Colors.success("\n✅ Generated commit message:"))
	print(Colors.highlight(commit_msg))
	print()
	
	edit = input(Colors.dim("Edit before commit? [y/N]: ")).strip().lower() == 'y'
	if edit:
		print(Colors.dim("Enter new commit message. End with an empty line:"))
		lines = []
		while True:
			line = input()
			if not line.strip():
				break
			lines.append(line)
		commit_msg = '\n'.join(lines) if lines else commit_msg
	
	commit(commit_msg)
	print(Colors.success("✅ Committed successfully."))
	
	if args.push:
		push()
		print(Colors.success("🚀 Pushed to remote."))
	else:
		print(Colors.dim("💡 Use --push flag to automatically push after commit."))

if __name__ == "__main__":
	main()

File: src/cli/commit.rs
use clap::Args;
use dialoguer::{Select, theme::SimpleTheme};
use regex::Regex;
use std::io;

use crate::config::AppConfig;
use crate::services::ai::{AiClient, Message};
use crate::utils::cleanup::clean_commit_message;
use crate::utils::colors::Colors;
use crate::utils::errors::AppResult;
use crate::utils::git::{commit, get_branch, get_diff, push, stage_all};

#[derive(Args, Debug, Clone)]
pub struct CommitArgs {
    #[arg(long, value_parser = ["detailed", "one-line"], help = "Commit message format")]
    pub format: Option<String>,
    #[arg(long, help = "Provider to use (overrides config)")]
    pub provider: Option<String>,
    #[arg(long, help = "Model to use (overrides config)")]
    pub model: Option<String>,
    #[arg(long, help = "Push after commit")]
    pub push: bool,
    #[arg(long, help = "Skip git hooks (pass --no-verify to git commit)")]
    pub no_verify: bool,
}

pub fn run(mut config: AppConfig, args: CommitArgs) -> AppResult<()> {
    if let Some(format) = &args.format {
        config.set_commit_format(format);
        config.save()?;
    }

    let provider = args.provider.clone().unwrap_or_else(|| config.provider());
    config.ensure_provider_ready(&provider)?;
    let ai = AiClient::new(config.clone())?;

    stage_all()?;
    let diff = get_diff(true, None)?;
    if diff.trim().is_empty() {
        println!("{}", Colors::info("[skip] no staged changes to commit"));
        return Ok(());
    }

    let branch = get_branch()?;
    let prefix = get_ticket_prefix(&branch);
    let short = config.commit_format() == "one-line";

    let system_msg = "You are an expert Git commit assistant. When responding, return only the commit message text itself—no extra explanation, quotes, or formatting. \
First, inspect the current Git branch name. If it begins with a ticket code matching the pattern LETTERS-DIGITS (for example ABC-123 or EL-2024), capture that exact code and place it at the very start of your message, followed by a colon and a space. If no ticket code is present, do not include any prefix. \
Next, identify the primary change or task implied by the branch name and present it as the first action in your commit message. Then, describe any secondary updates, fixes, or refactoring included in this commit. \
Use a natural, professional tone that reads like a teammate clearly explaining the work you’ve done. Use bullet points to separate multiple actions if they exist.";

    let user_msg = format!(
        "Branch: {branch}\nWrite a {} commit message for these changes:\n\n{diff}",
        if short {
            "one-line"
        } else {
            "detailed, human-friendly"
        }
    );

    println!("{}", Colors::header("[ai] generating commit message"));
    let messages = vec![
        Message {
            role: "system".to_string(),
            content: system_msg.to_string(),
        },
        Message {
            role: "user".to_string(),
            content: user_msg.clone(),
        },
    ];

    let mut commit_msg =
        ai.generate(&provider, &user_msg, Some(&messages), args.model.as_deref())?;
    commit_msg = clean_commit_message(&commit_msg);

    if !prefix.is_empty() && !commit_msg.starts_with(&prefix) {
        commit_msg = format!("{prefix}: {commit_msg}");
    }

    println!("{}", Colors::success("\n[ok] generated commit message"));
    println!("{}", Colors::highlight(&commit_msg));
    println!();

    if select_edit_action()? {
        println!(
            "{}",
            Colors::dim("Enter new commit message. End with an empty line:")
        );
        let mut lines = Vec::new();
        loop {
            let mut line = String::new();
            io::stdin().read_line(&mut line)?;
            let trimmed = line.trim_end();
            if trimmed.is_empty() {
                break;
            }
            lines.push(trimmed.to_string());
        }
        if !lines.is_empty() {
            commit_msg = lines.join("\n");
        }
    }

    commit(&commit_msg, args.no_verify)?;
    println!("{}", Colors::success("[ok] committed successfully"));

    if args.push {
        push()?;
        println!("{}", Colors::success("[ok] pushed to remote"));
    } else {
        println!(
            "{}",
            Colors::dim("[hint] use --push to push after committing")
        );
    }

    Ok(())
}

fn get_ticket_prefix(branch: &str) -> String {
    let re = Regex::new(r"^([A-Za-z]+-\d+)").unwrap();
    re.captures(branch)
        .and_then(|caps| caps.get(1).map(|m| m.as_str().to_string()))
        .unwrap_or_default()
}

fn select_edit_action() -> AppResult<bool> {
    let actions = ["commit generated message", "edit before commit"];
    let idx = Select::with_theme(&SimpleTheme)
        .with_prompt("next action")
        .items(actions)
        .default(0)
        .interact()?;
    Ok(idx == 1)
}
